UIP-0027.md

```diff
@@ -0,0 +1,88 @@
+# UIP-27: Transaction service fields
+
+```
+Author:   Dima Saveliev <dima@thirdhash.com>
+Status:   Draft
+Created:  2019-05-07
+```
+
+## Abstract
+
+  This document describes protocol changes in the transaction structure:
+the `nVersion` field is going to be replaced by four separate service fields.
+
+
+## Motivation
+
+  UIP-0018 introduced a transaction type concept and the corresponding field was added
+as part of already existing `nVersion` field. That was achieved via some chunks
+of bit arithmetic and was a kind of ad-hoc hack.
+
+  By implementing separate service fields we solve the following problems:
+  
+  1. Clean-up the logic behind the `type` and `version` fields.
+  2. Minimize the size of `type` and `version` fields.
+  3. Preserve compatibility with Bitcoin codebase.
```

Comment:
How can we preserve compatibility with the Bitcoin codebase if we are changing the field? That might need a bit more explanation. The current solution (before this UIP is implemented) isn't less compatible than the one proposed here, isn't it? Mabye this point is less a problem solution but more a constraint we want to keep.

Reply:
@cornelius , Shouldn't the "how" be further explained in the "Rationale" section?

Reply:
I think the "how" belongs to the specification section and the "why" to the rationale section, and I think they both have the information which is needed. Here in the motivation section I would outline the reason and requirements for the change. It's probably just a matter of wording and maybe also adding a pointer to the rationale which comes later.

Reply:
Yeah I don't think we can keep compatibility this way, we are changing the tx datagram...

Reply:
Well, I wasn't accurate with this definition, my fault.
By compatibility, I mean a byte amount and order (since the actual `version` value doesn't exceed 1 byte, we still would have it in the right position), which mimics the old `nVersion` field (and that helps us to use old tx stubs, keep assumptions about tx size, etc.)  - just what we have right now with `nVersion` overloading.

Reply:
I changed the wording of the last clause